app/jobs/client_login_request_job.rb

```diff
@@ -0,0 +1,34 @@
+class ClientLoginRequestJob < ApplicationJob
+  def perform(email_address: nil, phone_number: nil)
+    raise StandardError("No contact info provided") unless email_address.present? || phone_number.present?
+
+
```

Comment:
I remember we talked about two different models we could use as a key for rate-limit data: The `Client`, or the contact point (phone number/email address). I recall that you were in favor of using the `Client` as the rate limiting key, since most of the time, it'd be 1 client, and the worst you could do is lock out another client who shares your contact info, which would be annoying, but we'd get past it. That could be fine.

As I understand it, the rate limit here would be primarily about not annoying our users, secondarily about not getting us blocked as spam from contacting specific users, and not fundamentally about authentication security.

For that situation, you can imagine a max of (say) 12 login attempts per day, past which we refuse to take action. If that's adequate, then you could imagine a table like `ClientLoginRequest` which has:

- client_id
- number
- period_start

`period_start` would be the current time rounded to the start of the day, Calling it `period_start` implies we might change the time period, but we could call it `day_start` and that would be fine, too.

You could `find_or_create` the `client_id, period_start` pair, and then increment the number, and if it's too high, refuse to take action.

This would create some database garbage at a rate of 1 record per client per day. We would probably be fine leaving this database garbage in place. If we want to clear the garbage, an opportune time to clear it is after reading each client.

Reply:
In the case that the rate limit is exceeded, what exactly do we do? I guess we can send a final message to the client saying that we rate limited them but please try back tomorrow, or email our support?